crates/client/db/src/rocksdb_snapshot.rs

```diff
@@ -0,0 +1,169 @@
+use librocksdb_sys as ffi;
+use rocksdb::{AsColumnFamilyRef, DBAccess, DBPinnableSlice, Error, ReadOptions};
+use std::sync::Arc;
+
+/// A copy of [`rocksdb::SnapshotWithThreadMode`] with an Arc<DB> instead of an &'_ DB reference
+pub struct SnapshotWithDBArc<D: DBAccess> {
+    db: Arc<D>,
+    // The Database needs to outlive the snapshot, and be created by the supplied `db`.
+    inner: *const ffi::rocksdb_snapshot_t,
+}
+
+#[allow(dead_code)]
+impl<D: DBAccess> SnapshotWithDBArc<D> {
+    /// Creates a new `SnapshotWithDBArc` of the database `db`.
+    pub fn new(db: Arc<D>) -> Self {
+        let snapshot = unsafe { db.create_snapshot() };
+        Self { db, inner: snapshot }
+    }
+
+    /// Returns the bytes associated with a key value with default read options.
+    pub fn get<K: AsRef<[u8]>>(&self, key: K) -> Result<Option<Vec<u8>>, Error> {
+        let readopts = ReadOptions::default();
+        self.get_opt(key, readopts)
+    }
+
+    /// Returns the bytes associated with a key value and given column family with default read
+    /// options.
+    pub fn get_cf<K: AsRef<[u8]>>(&self, cf: &impl AsColumnFamilyRef, key: K) -> Result<Option<Vec<u8>>, Error> {
+        let readopts = ReadOptions::default();
+        self.get_cf_opt(cf, key.as_ref(), readopts)
+    }
+
+    /// Returns the bytes associated with a key value and given read options.
+    pub fn get_opt<K: AsRef<[u8]>>(&self, key: K, mut readopts: ReadOptions) -> Result<Option<Vec<u8>>, Error> {
+        // Safety: the snapshot originates from the `db`, and it is not dropped during the lifetime of the `readopts` variable.
+        unsafe {
+            readopts.set_raw_snapshot(self.inner);
+        }
```

Comment:
Extract this out into a private function. It's repeated several times, and would keep the unsafe call into one place.

Reply:
actually i don't agree with that
that would technically mean that this private function should be unsafe to call, since it can't uphold the readopts lifetime requirement
so we would still need an unsafe block at all of these places and a safety comment to go along with them

Reply:
actually there's a smarter way to avoid repeating that safety comment a bunch of times, it looks like this

```rs
    // This function allows us to not repeat the unsafe block a bunch of times. It takes ownership of the
    // readoptions to ensure its lifetime is stictly smaller than self.
    fn readopts_with_raw_snapshot<R>(&self, mut readopts: ReadOptions, f: impl FnOnce(&ReadOptions) -> R) -> R {
        // Safety: the snapshot originates from the `db`, and it is not dropped during the lifetime of the `readopts` variable.
        unsafe {
            readopts.set_raw_snapshot(self.inner);
        }

        f(&readopts)
    }
```
that way the function is safe